src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs

```diff
@@ -0,0 +1,420 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO.Pipelines;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using static Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    internal class MsQuicStream : TransportConnection, IUnidirectionalStreamFeature
+    {
+        private Task _processingTask;
+        private MsQuicConnection _connection;
+        private readonly CancellationTokenSource _streamClosedTokenSource = new CancellationTokenSource();
+        private IMsQuicTrace _trace;
+        private bool _disposed;
+        private IntPtr _nativeObjPtr;
+        private GCHandle _handle;
+        private readonly IntPtr _unmanagedFnPtrForNativeCallback;
+
+        internal ResettableCompletionSource _resettableCts;
+        private MemoryHandle[] _bufferArrays;
+        private GCHandle _sendBuffer;
+
+        public MsQuicStream(MsQuicApi registration, MsQuicConnection connection, MsQuicTransportContext context, QUIC_STREAM_OPEN_FLAG flags, IntPtr nativeObjPtr)
+        {
+            Debug.Assert(connection != null);
+
+            Registration = registration;
+            _nativeObjPtr = nativeObjPtr;
+            _unmanagedFnPtrForNativeCallback = Marshal.GetFunctionPointerForDelegate((StreamCallbackDelegate)NativeCallbackHandler);
+
+            _connection = connection;
+            MemoryPool = context.Options.MemoryPoolFactory();
+            _trace = context.Log;
+
+            ConnectionClosed = _streamClosedTokenSource.Token;
+
+            var maxReadBufferSize = context.Options.MaxReadBufferSize.Value;
+            var maxWriteBufferSize = context.Options.MaxWriteBufferSize.Value;
+            _resettableCts = new ResettableCompletionSource(this);
+
+            // TODO should we allow these PipeScheduler to be configurable here?
+            var inputOptions = new PipeOptions(MemoryPool, PipeScheduler.ThreadPool, PipeScheduler.Inline, maxReadBufferSize, maxReadBufferSize / 2, useSynchronizationContext: false);
+            var outputOptions = new PipeOptions(MemoryPool, PipeScheduler.Inline, PipeScheduler.ThreadPool, maxWriteBufferSize, maxWriteBufferSize / 2, useSynchronizationContext: false);
+
+            var pair = DuplexPipe.CreateConnectionPair(inputOptions, outputOptions);
+
+            if (flags.HasFlag(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL))
+            {
+                Features.Set<IUnidirectionalStreamFeature>(this);
+            }
+
+            // TODO populate the ITlsConnectionFeature (requires client certs). 
```

Comment:
We'll also need ITlsHandshakeFeature, though that makes more sense at the connection level.

Reply:
I'll leave a todo in the connection